src/v/cloud_topics/level_one/frontend_reader/reader.cc

```diff
@@ -0,0 +1,415 @@
+          "size {}): {}",
+          oid,
+          extent.position,
+          object_size,
+          stream_fut.get_exception());
+        co_return std::unexpected(l1::io::errc::file_io_error);
+    }
+    auto stream_result = stream_fut.get();
+    if (!stream_result.has_value()) {
+        co_return std::unexpected(stream_result.error());
+    }
+
+    auto& stream = stream_result.value();
+
+    auto footer_fut = co_await ss::coroutine::as_future(
+      read_iobuf_exactly(stream, footer_total_size));
+    if (footer_fut.failed()) {
+        vlog(
+          cd_log.error,
+          "Exception reading footer from object {} (pos {} object size {}): {}",
+          oid,
+          extent.position,
+          object_size,
+          footer_fut.get_exception());
+        co_await stream.close();
+        co_return std::unexpected(l1::io::errc::file_io_error);
+    }
+
+    iobuf footer_buf = footer_fut.get();
+    co_await stream.close();
+
+    // Parse the footer - we have the complete footer so this should succeed.
+    auto footer_result = co_await l1::footer::read(std::move(footer_buf));
+
+    if (!std::holds_alternative<l1::footer>(footer_result)) {
+        vlog(
+          cd_log.error,
+          "Failed to parse footer from object {} despite reading complete "
+          "footer (pos {} object size {})",
+          oid,
+          extent.position,
+          object_size);
+        co_return std::unexpected(l1::io::errc::file_io_error);
+    }
+
+    co_return std::get<l1::footer>(std::move(footer_result));
+}
+
+ss::future<> level_one_log_reader_impl::materialize_batches(
+  model::timeout_clock::time_point /*deadline*/) {
+    // Could be EOS because there are no more objects, but
+    // there should never be materialized batches remaining.
+    vassert(
+      _batches.empty(),
+      "Materialize batches called with batches already materialized");
+
+    if (_state == state::end_of_stream) {
+        co_return;
+    }
+
+    vassert(
+      _state == state::ready,
+      "Invalid state to materialize batches: {}",
+      std::to_underlying(_state));
+
+    // I wish I had ADTs to enforce these invariants...
+    vassert(
+      _current_obj.has_value(),
+      "Expected to have current object in ready state");
+
+    size_t start_pos = _current_obj->footer.file_position_before_kafka_offset(
+      _tidp, _config.start_offset);
+    if (start_pos == l1::footer::npos) {
+        // Perhaps this object spans offsets in the metastore but has
+        // no data because of compaction.
+        vlog(
+          cd_log.debug,
+          "No data for {} ({}) in object {}: materializing 0 batches",
+          _ntp,
+          _tidp,
+          _current_obj->oid);
+        _state = state::materialized;
+        co_return;
+    }
+
+    l1::object_extent extent{
+      .id = _current_obj->oid,
+      .position = start_pos,
+      .size = L1_max_bytes_per_object_fetch,
```

Comment:
Thinking more, we'll probably think we get corrupted data in the L1 reader if we're reading a partial object because it was truncated in the middle. I think we *have* to pick boundaries that are defined in the footer, or we need to make modifications to the l1 object reader to understand it could be called with partial data.

My heavy preference is to just always align on boundaries for now, supporting partial data in the reader will complicate the code a bunch to distinguish this case between real corruption.

How do the tests not catch this? Are all the batches super small?